include/subsystems/odometry/odometry_nwheel.h

```diff
@@ -0,0 +1,271 @@
+   *
+   * @return the robot's updated position
+   */
+  pose_t update() override {
+    Eigen::Vector<double, WHEELS> radian_deltas;
+
+    for (int i = 0; i < WHEELS; i++) {
+      double angle_rad = encoders[i].position(rev) * M_PI * 2;
+      radian_deltas(i) = angle_rad - old_wheel_angles[i];
+
+      old_wheel_angles[i] = angle_rad;
+    }
+
+    pose_t updated_pos = calculate_new_pos(radian_deltas, current_pos);
+
+    // if we do not pass in an IMU we use the wheels for rotation
+    if (imu != nullptr) {
+      angle = 0;
+      // Translate "0 forward and clockwise positive" to "CCW positive and radians"
+      angle = -imu->rotation(vex::rotationUnits::rev) * 2 * M_PI;
+      // Offset the angle, if we've done a set_position
+      angle += angle_offset;
+    }
+
+    static pose_t last_pos = updated_pos;
+    static double last_speed = 0;
+    static double last_ang_speed = 0;
+    static timer tmr;
+
+    double speed_local = 0;
+    double accel_local = 0;
+    double ang_speed_local = 0;
+    double ang_accel_local = 0;
+    bool update_vel_accel = tmr.time(sec) > 0.1;
+
+    // This loop runs too fast. Only check at LEAST every 1/10th sec
+    if (update_vel_accel) {
+      // Calculate robot velocity
+      speed_local = pos_diff(updated_pos, last_pos) / tmr.time(sec);
+
+      // Calculate robot acceleration
+      accel_local = (speed_local - last_speed) / tmr.time(sec);
+
+      // Calculate robot angular velocity (deg/sec)
+      ang_speed_local = smallest_angle(updated_pos.rot, last_pos.rot) / tmr.time(sec);
+
+      // Calculate robot angular acceleration (deg/sec^2)
+      ang_accel_local = (ang_speed_local - last_ang_speed) / tmr.time(sec);
+
+      tmr.reset();
+      last_pos = updated_pos;
+      last_speed = speed_local;
+      last_ang_speed = ang_speed_local;
+    }
+
+    this->current_pos = updated_pos;
+    if (update_vel_accel) {
+      this->speed = speed_local;
+      this->accel = accel_local;
+      this->ang_speed_deg = ang_speed_local;
+      this->ang_accel_deg = ang_accel_local;
+    }
+
+    if (imu != nullptr) {
+      old_heading = angle;
+    }
+
+    return current_pos;
+  }
+
+  /**
+   * Resets the position and rotational data to the input.
+   */
+  void set_position(const pose_t &newpos) override {
+    mut.lock();
+    rotation_offset = newpos.rot - (current_pos.rot - rotation_offset);
+    mut.unlock();
+
+    OdometryBase::set_position(newpos);
+  }
+
+private:
+  /**
+   * Applies a twist (pose delta) to a pose by including first order dynamics of heading.
+   * Can be thought of as applying a twist as following an arc rather than a straight line.
+   *
+   * https://file.tavsys.net/control/controls-engineering-in-frc.pdf#section.10.2
+   *
+   * @param old_pose  The pose to which the twist will be applied
+   * @param twist     The twist, represents a pose delta
+   */
+  pose_t pose_exponential(const Eigen::Vector3d old_pose, const Eigen::Vector3d twist) {
```

Comment:
can you move this to an odometry_base static method or math utils or something
where its not explicilty tied to n pod it would be nice to have it available elsewhere

Reply:
for now I'm gonna put this in odometry_base, when I make pose2d I'll move it in there (and also probably make a pose_logarithm function as well